File: java/AggregateOperator.java
import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;

import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.api.java.function.Function2;
import org.apache.spark.api.java.function.VoidFunction;

/**
 * aggregate算子演示
 * 会根据定义的第一个匿名函数对每个分区的数据进行处理
 * 会根据所有分区的返回值进行第二个匿名函数的操作
 * 该算子在分区数为1时不会调用第二个匿名函数
 */
public class AggregateOperator {

	public static void main(String[] args) {
		SparkConf sparkConf = new SparkConf().setAppName("AggregateOperator").setMaster("local[10]");
		JavaSparkContext sc = new JavaSparkContext(sparkConf);
		List<Integer> dataList =  new ArrayList<Integer>();
		dataList.add(1);
		dataList.add(2);
		dataList.add(3);
		dataList.add(4);
		dataList.add(5);
		dataList.add(6);

		JavaRDD<Integer> dataRdd = sc.parallelize(dataList,3);

		JavaRDD<String> nameWithPartitonIndex = dataRdd.mapPartitionsWithIndex(new Function2<Integer, Iterator<Integer>, Iterator<String>>() {

			public Iterator<String> call(Integer index, Iterator<Integer> iterator) throws Exception {
				List<String> list = new ArrayList<String>();
				while(iterator.hasNext()){
					Integer name = iterator.next();
					String result = index + " : " + name;
					list.add(result);
				}
				return list.iterator();
			}
		}, true);
		 
		nameWithPartitonIndex.foreach(new VoidFunction<String>() {

			public void call(String result) throws Exception {
				System.out.println(result);
			}
		});
		 
		Integer result = dataRdd.aggregate(2,new Function2<Integer,Integer,Integer>() {

			public Integer call(Integer v1, Integer v2) throws Exception {
				System.out.println("seq:"+v1+"==="+v2);
				return Math.min(v1, v2);
			}
		},new Function2<Integer,Integer,Integer>() {

			public Integer call(Integer v1, Integer v2) throws Exception {
				System.out.println("comb:"+v1+"==="+v2);
				return v1+v2;
			}

		});
		
		System.out.println(result);
	}
}

File: scala/AggregateOperator.scala
import org.apache.spark.{SparkContext, SparkConf}

/**
  * aggregateByKey算子演示
  * 会根据key进行聚合
  * 会根据定义的seq函数比较应该返回的值并对所有返回的值进行comb操作
  * 该算子需要注意本地执行和服务器执行会有不同，保险起见本地执行设置为*
  */
object AggregateOperator {

  def main(args: Array[String]) {

    // 设置环境参数
    val conf = new SparkConf().setAppName("AggregateOperator").setMaster("local[*]")

    // 初始化执行环境
    val sc = new SparkContext(conf)

    // 初始化并对数据进行分区操作
    val dataRdd = sc.parallelize(List((1,3),(1,2),(1,4),(2,3)))

    // 定义聚合操作
    def comb(a:Int,b:Int):Int = {
      println("comb:" + a + "\t" + b)
      a + b
    }

    // 定义比较操作
    def seq(a:Int,b:Int):Int = {
      println("seq:" + a + "\t" + b)
      math.max(a,b)
    }

    // 使用3作为中间值，并进行collect(提取所有数据)的action操作，最后进行foreach操作
    dataRdd.aggregateByKey(3)(seq,comb).collect().foreach(println)

    // 输出结果中，会存在
    // seq:3	3
    // seq:3	4
    // seq:3	3
    // seq:3 2
    // comb:3	3
    // comb:6	4
    // (1,10)
    // (2,3)
    // 会调用seq和comb对数据做处理
  }
}
